verification/src/solidity/compiler_fetcher.rs

```diff
@@ -104,15 +108,97 @@ impl TryFrom<(json::CompilerInfo, &Url)> for CompilerInfo {
     }
 }
 
+#[derive(Default)]
+struct RefreshableCompilerVersions {
+    versions: Arc<RwLock<CompilerVersions>>,
+}
+
+impl From<CompilerVersions> for RefreshableCompilerVersions {
+    fn from(versions: CompilerVersions) -> Self {
+        Self {
+            versions: Arc::new(RwLock::new(versions)),
+        }
+    }
+}
+
 #[derive(Default)]
 pub struct CompilerFetcher {
-    releases: Releases,
+    compiler_versions: RefreshableCompilerVersions,
     folder: PathBuf,
 }
 
 impl CompilerFetcher {
-    pub async fn new(releases: Releases, folder: PathBuf) -> Self {
-        Self { releases, folder }
+    pub async fn new(
+        versions_list_url: Url,
+        refresh_cron_schedule: Option<&str>,
+        folder: PathBuf,
+    ) -> anyhow::Result<Self> {
+        let compiler_versions = CompilerVersions::fetch_from_url(&versions_list_url)
+            .await
+            .map_err(anyhow::Error::msg)?;
+        let compiler_versions = RefreshableCompilerVersions::from(compiler_versions);
+        if let Some(cron_schedule) = refresh_cron_schedule {
+            compiler_versions
+                .spawn_refresh_job(versions_list_url, cron_schedule)
+                .map_err(anyhow::Error::msg)?;
+        }
+        Ok(Self {
+            compiler_versions,
+            folder,
+        })
+    }
+}
+
+impl RefreshableCompilerVersions {
+    fn spawn_refresh_job(
+        &self,
+        versions_list_url: Url,
+        cron_schedule: &str,
+    ) -> Result<(), JobSchedulerError> {
+        log::info!("spawn version refresh job with schedule {}", cron_schedule);
+        let versions = Arc::clone(&self.versions);
+
+        let job = Job::new_async(cron_schedule, move |_, _| {
+            let versions_list_url = versions_list_url.clone();
+            let versions = Arc::clone(&versions);
+
+            Box::pin(async move {
+                let refresh_result =
+                    RefreshableCompilerVersions::refresh_versions(versions, &versions_list_url)
+                        .await;
+                if let Err(err) = refresh_result {
+                    log::error!("error during version refresh: {}", err);
+                };
+            })
+        })?;
+        let scheduler = JobScheduler::new()?;
+        scheduler.add(job)?;
+        scheduler.start()
+    }
+
+    async fn refresh_versions(
+        versions: Arc<RwLock<CompilerVersions>>,
+        versions_list_url: &Url,
+    ) -> anyhow::Result<()> {
+        let fetched_versions = CompilerVersions::fetch_from_url(versions_list_url)
+            .await
+            .map_err(anyhow::Error::msg)?;
+        let need_to_update = {
+            let versions_lock = versions.read().await;
```

Comment:
about naming the variables with mutexes
usually it's something like
```rust
let variable_lock: Mutex<Type>
let variable_guard: MutexGuard<Type>
let variable: &Type
```
or you can just name the `variable_guard` as `variable`, if the name is free
let's rename them everywhere